### PR TITLE
*: update rust-toolchain

### LINKED ISSUE #15581: update rust toolchain
## Development Task

The currently rust-toolchain `nightly-2022-11-15` is too old we should update it to a recent version.

### PR DESCRIPTION
**Pull request by glorv**
### What is changed and how it works?

Issue Number: Close #15581

What's Changed:

This PR update the rust-toolchain to nightly-2023-08-15 and fix most clippy error and warnings. There are still some clippy warnings that are not fixed either because it is false positive or hard to fix.

```
Update rust-toolchain to nightly-2023-08-15 and fix build&clippy errors
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by overvenus — APPROVED** [review]

**Review by Connor1996 — APPROVED** [review]
LGTM

File: components/raftstore-v2/src/lib.rs [hunk 1]
```diff
@@ -27,7 +27,8 @@
 #![feature(box_into_inner)]
 #![feature(assert_matches)]
 #![feature(option_get_or_insert_default)]
-#[allow(clippy::needless_pass_by_ref_mut)]
+#![allow(clippy::needless_pass_by_ref_mut)]
```
Comment (overvenus) [inline components/raftstore-v2/src/lib.rs:30]: You can add it to Makefile.

Comment (glorv) [reply components/raftstore-v2/src/lib.rs:30]: But I only want to allow this for some components temporarily. Some are not easy to fix, so we can fix them in following PRs.

**Pull request merged**